### ISSUE TITLE
any option to have [GetComponent] not override user set reference?

### ISSUE THREAD

**Issue by laurentopia** [TylerTemp/SaintsField#72]
I'd like to use [GetComponent] to auto set a default value that can be overridden by the user.
But GetComponent overrides user value.
Is there a way around that?

**Comment by TylerTemp** [comment]
Hi,

this is now possible in [3.4.0](https://github.com/TylerTemp/SaintsField/releases/tag/3.4.0):

```csharp
[GetComponent(EXP.Silent | EXP.NoPicker)] public Dummy issue72Dummy;
[GetComponent(EXP.Silent | EXP.NoPicker | EXP.NoResignButton)] public Dummy issue72DummySimple;
```

![image](https://github.com/user-attachments/assets/8cf3a510-268f-42b4-aad9-7fb5a69b63f8)

The config is kinda tedious. It might be optimized on a later version.

Note: This update might contain bugs.

**Comment by laurentopia** [comment]
is it X or the cycling button? in the other post it was an X so maybe I don't understand its purpose

**Comment by TylerTemp** [comment]
Hi,

"X" means the value is manually signed, but the `Get*` attribute actually find nothing, so a "x" will allow you to remove the value.

"Reload" means the value is manually signed, but the `Get*` attibute find a value which is different from the current value, so a "Reload" will allow you to sign to a "correct" value, rather than the "current" value.

And `EXP.NoResignButton` will completely hide these buttons.

**Comment by laurentopia** [comment]
this isn't working

https://github.com/user-attachments/assets/864d5d17-8540-4703-956e-b63188342906

![image](https://github.com/user-attachments/assets/b9bb4fa8-9410-49c8-bdac-33afe567b87e)

**Comment by TylerTemp** [comment]
Hi,

To use this feature, you need to at lease set `EXP.Silent`:

```csharp
// simple
[GetComponentInChildren(EXP.Silent)] public MCMineable source;
// no picker button, no resign/remove button
[GetComponentInChildren(EXP.Silent | EXP.NoPicker | EXP.NoResignButton)] public MCMineable sourcePlain;
```

https://github.com/user-attachments/assets/84c383d2-78bd-4b3f-945b-6bd03130e928


---------------------

I do plan to make a config function to set default config globally for many attributes. But that will be much later, after I finish some important features first.

**Comment by laurentopia** [comment]
Ok. Why silent? Seems like a good default behavior.

**Comment by TylerTemp** [comment]
Well, I can not give a good reason.

When I design `GetComponentByPath`, which was originally designed to be the parent of all the auto-getter attributes, it behaves exactly like `Silent`. I'd like to think of it as a good design.

When I refactored it into `GetByXPath`, it took way too much time than I expected. When on its final touch, I actually lost track of which should be the default behavior. 

When I'm working in the company, one thing I need to do constantly is: to re-bind all the lost references after a code merge. Yep, we decide not to merge the Unity scene because it's just not possible: Unity has a so-called `yaml-merger`, and it sucks. So whenever I merge the scenes, I just throw away my scenes and manually add everything back into the remote one. After manually adding the object & script, then is the magic part: because the auto-getter attributes will aggressively sign the value to the "found" value, all my setup will be correct, with no need to manually click the "re-sign button".

So, with not enough thought, I picked that aggressive behavior as the default. And I regret it...

Yep, that's the story. That's also why I plan to support a config file to allow to change of the default behavior globally (like what `DOTween` does).

**Comment by laurentopia** [comment]
oh boy, it's like thinking of a name for a method ...

**Comment by TylerTemp** [comment]
> Ok. Why silent? Seems like a good default behavior.

Hi,

Since [3.4.11](https://github.com/TylerTemp/SaintsField/releases/tag/3.4.11) you can use `Window/Saints/Create or Edit SaintsField Config` to tweak the config of the default behavior of auto getters.

**Comment by laurentopia** [comment]
Cool, I was just running into default collision!

**Comment by TylerTemp** [comment]
> just updated to 3.4.11 and I'm running into this new bug. ![image](https://private-user-images.githubusercontent.com/26075468/385634345-7df84efe-8f8b-437c-a5ca-bfa899ed1836.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzE0ODgyODYsIm5iZiI6MTczMTQ4Nzk4NiwicGF0aCI6Ii8yNjA3NTQ2OC8zODU2MzQzNDUtN2RmODRlZmUtOGY4Yi00MzdjLWE1Y2EtYmZhODk5ZWQxODM2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTEzVDA4NTMwNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTVhNjRiNDQ4OTEyYTRjNjU2OTIyMGI3YzA0OTVmYmEyOTQxMmNkMDA5NmY0MjkyMWE0MGE0ZWZhZTViYjRmM2ImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.g088BxfKHIt7uzn_8VEqnU-9z25mCthRhuYlARGSt3s) Maybe this accessor was added after 2020

Hi,

Can you copy paste the error? Looks like some issue with reading the config file

**Comment by TylerTemp** [comment]
... wait it did has an issue...

**Comment by TylerTemp** [comment]
Oh holy fuck... your version is indeed a bit low that does not have ManagedReference yet

I'll fix it soon

**Comment by TylerTemp** [comment]
Sorry for the trouble, it's been fixed in [3.4.12](https://github.com/TylerTemp/SaintsField/releases/tag/3.4.12)

**Comment by laurentopia** [comment]
No worry, I needed to go to bed anyway :D
